### PR TITLE
Assigment3

### PR DESCRIPTION
**Pull request by antoniomcoelho**

### CODE REVIEW

File: life_expectancy/tests/test_load_save.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    pd.testing.assert_frame_equal(
+        load_data_actual, load_data_expected
+    )
+
+
+def test_save_data(pt_life_expectancy_expected):
+    """Run the `save_data` function and compare the output to the expected output"""
+    save_data_expected = pt_life_expectancy_expected
+    with mock.patch.object(save_data_expected, "to_csv") as to_csv_mock:
```
Comment (anaritarc) [inline life_expectancy/tests/test_load_save.py:24]: this is awesome :star-struck:

File: life_expectancy/tests/test_load_save.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+"""Tests for the cleaning module"""
+import pandas as pd
+import unittest.mock as mock
+
+from life_expectancy.cleaning import clean_data
```
Comment (anaritarc) [inline life_expectancy/tests/test_load_save.py:5]: clean_data is not used here, I would remove it 😄

File: life_expectancy/tests/test_load_save.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from life_expectancy.load_save import load_data, save_data
+from . import INIT_DIR, FIXTURES_DIR, OUTPUT_DIR
```
Comment (anaritarc) [inline life_expectancy/tests/test_load_save.py:7]: you have called init_dir but it isn't used, I would also remove it

File: life_expectancy/tests/test_load_save.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def test_load_data(load_data_expected):
+    """Run the `load_data` function and compare the output to the expected output"""
+
+    load_data_actual = load_data()
+    #load_data_expected = pd.read_csv( FIXTURES_DIR / "eu_life_expectancy_raw.tsv", sep="\t")
```
[the remark this answers is not in the document]
Comment (anaritarc) [reply life_expectancy/tests/test_load_save.py:14]: Also I had the recommendation to also use patch in the load data, I believe you could also try it ❓

**Review by anaritarc — CHANGES_REQUESTED** [review]

File: life_expectancy/load_save.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+''' Clean data code'''
+import argparse
+from pathlib import Path # pylint: disable=import-error
```
Comment (MrCordeiro) [inline life_expectancy/load_save.py:3]:
```suggestion
from pathlib import Path
```

Algum outra cena pode estar a causar este import error

File: life_expectancy/load_save.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+''' Clean data code'''
+import argparse
```
Comment (MrCordeiro) [inline life_expectancy/load_save.py:2]: Validar se não apareceu um "unused-import"

File: life_expectancy/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+''' Pipeline for clean data code'''
+import argparse
+from life_expectancy.load_save import load_data, save_data
+from life_expectancy.cleaning import clean_data
+
+
+
+def add_region_user():
```
Comment (MrCordeiro) [inline life_expectancy/pipeline.py:8]:
```suggestion
def _add_region_user() -> str:
```

File: life_expectancy/tests/__init__.py [hunk 8]
```diff
@@ -4,3 +4,4 @@
 
 FIXTURES_DIR = Path(__file__).parent / "fixtures"
 OUTPUT_DIR = Path(__file__).parent.parent / "data"
+INIT_DIR = Path(__file__).parent.parent
```
Comment (MrCordeiro) [inline life_expectancy/tests/__init__.py:7]: Validar se está a ser utilizado (nome generico)

File: life_expectancy/tests/conftest.py [hunk 9]
```diff
@@ -13,11 +13,17 @@ def run_before_and_after_tests() -> None:
     yield # this is where the testing happens
 
     # Teardown : fill with any logic you want
-    file_path = OUTPUT_DIR / "pt_life_expectancy.csv"
-    file_path.unlink(missing_ok=True)
+    #file_path = OUTPUT_DIR / "pt_life_expectancy.csv"
+    #file_path.unlink(missing_ok=True)
```
Comment (MrCordeiro) [inline life_expectancy/tests/conftest.py:17]: Deeeeeelete

File: life_expectancy/tests/test_load_save.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import unittest.mock as mock
+
+from life_expectancy.cleaning import clean_data
+from life_expectancy.load_save import load_data, save_data
+from . import INIT_DIR, FIXTURES_DIR, OUTPUT_DIR
+
+
+def test_load_data(load_data_expected):
+    """Run the `load_data` function and compare the output to the expected output"""
+
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_load_save.py:12]:
```suggestion
from unittest.mock import patch


@patch("life_expectancy.load_save.pd.read_table")
def test_load_data(mock_load, load_data_expected):
    """Run the `load_data` function and compare the output to the expected output"""
    
    mock_load.return_value = "INSERT FIXTURE"

```

File: life_expectancy/tests/test_load_save.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        load_data_actual, load_data_expected
+    )
+
+
+def test_save_data(pt_life_expectancy_expected):
+    """Run the `save_data` function and compare the output to the expected output"""
+    save_data_expected = pt_life_expectancy_expected
+    with mock.patch.object(save_data_expected, "to_csv") as to_csv_mock:
+        save_data(save_data_expected)
+        to_csv_mock.assert_called_with(OUTPUT_DIR / "pt_life_expectancy.csv", index=False)
```
[the remark this answers is not in the document]
Comment (MrCordeiro) [reply life_expectancy/tests/test_load_save.py:26]: É uma maneira bué boa!

File: life_expectancy/tests/test_load_save.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    pd.testing.assert_frame_equal(
+        load_data_actual, load_data_expected
+    )
+
+
+def test_save_data(pt_life_expectancy_expected):
+    """Run the `save_data` function and compare the output to the expected output"""
+    save_data_expected = pt_life_expectancy_expected
+    with mock.patch.object(save_data_expected, "to_csv") as to_csv_mock:
+        save_data(save_data_expected)
```
Comment (MrCordeiro) [inline life_expectancy/tests/test_load_save.py:25]:
```
        to_csv_mock.side_effect = print("I'm being saved!!)
```

**Review by MrCordeiro — CHANGES_REQUESTED** [review]

**Review by antoniomcoelho — COMMENTED** [review]
All changes done.

**Review by anaritarc — APPROVED** [review]
Awesome Antonio 💪🏼 🥳

File: life_expectancy/tests/test_load_save.py (hunk 10 above)
Comment (MrCordeiro) [reply life_expectancy/tests/test_load_save.py:12]: I don't see any changes here. Why is it marked as resolved?

File: life_expectancy/tests/test_load_save.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+def test_load_data(load_data_expected):
+    """Run the `load_data` function and compare the output to the expected output"""
+
+    load_data_actual = load_data()
+    #load_data_expected = pd.read_csv( FIXTURES_DIR / "eu_life_expectancy_raw.tsv", sep="\t")
+
+    pd.testing.assert_frame_equal(
+        load_data_actual, load_data_expected
+    )
```
[the remark this answers is not in the document]
Comment (MrCordeiro) [reply life_expectancy/tests/test_load_save.py:17]: This comment is a complement to the previous one.

**Review by MrCordeiro — CHANGES_REQUESTED** [review]
There were 2 comments that were marked as resolved, but that I couldn't find what changed.

The `test_load_data` is consuming data from the `data` directory, which we don't want (as it's our "database" in the context of this exercise)
Instead, you should patch the method that is doing the saving and replace it with a fixture.

File: life_expectancy/tests/test_load_save.py (hunk 10 above)
Comment (anaritarc) [reply life_expectancy/tests/test_load_save.py:12]: my bad sorry @MrCordeiro

Comment (MrCordeiro) [reply life_expectancy/tests/test_load_save.py:12]: Wait... haha...did you mark it as resolved, @anaritarc ?

Comment (anaritarc) [reply life_expectancy/tests/test_load_save.py:12]: I did :) I though it was done 🙈

**Review by MrCordeiro — APPROVED** [review]

**Pull request merged**